src/cascadia/TerminalControl/ScrollBarUpdater.cpp

```diff
@@ -0,0 +1,152 @@
+
+// Method Description:
+// - Apply the update to a scroll bar.
+// Arguments:
+// - scrollBar: the scroll bar
+// - modifyValue: whether the scroll bar's value should modified according to the update or not
+// Return Value:
+// - <none>
+void ScrollBarUpdate::Apply(ScrollBar scrollBar, bool modifyValue) const
+{
+    scrollBar.Maximum(NewMaximum);
+    scrollBar.Minimum(NewMinimum);
+    scrollBar.ViewportSize(NewViewportSize);
+
+    if (modifyValue)
+    {
+        scrollBar.Value(NewValue);
+    }
+}
+
+ScrollBarUpdater::ScrollBarUpdater(ScrollBar scrollBar) :
+    _scrollBar(scrollBar),
+    _throttle(std::chrono::milliseconds(ScrollRateLimitMilliseconds)),
+    _stillExists(std::make_shared<bool>(true))
+{
+}
+
+void ScrollBarUpdater::SetScrollBar(ScrollBar scrollBar)
+{
+    _scrollBar = scrollBar;
+}
+
+// Method Description:
+// - Prepare to update the scroll bar later (asynchronously).
+// Arguments:
+// - update: the update with the new scroll bar's settings
+// Return Value:
+// - <none>
+void ScrollBarUpdater::DoUpdate(ScrollBarUpdate update)
+{
+    {
+        std::lock_guard guard(_pendingUpdateMutex);
+
+        bool isUpdaterRunning = _pendingUpdate.has_value();
+        _pendingUpdate = { update };
+
+        if (isUpdaterRunning)
+        {
+            return;
+        }
+    }
+
+    _RunUpdater();
+}
+
+// Method Description:
+// - Call this from the scroll bar's value change event handler.
+// - It returns whether that event is coming from us or from an external
+//   source such as the user.
+// Arguments:
+// - <none>
+// Return Value:
+// - `true` if the current event is from us. `false` otherwise.
+bool ScrollBarUpdater::IsInternalUpdate() const
+{
+    return _isInternalUpdate;
+}
+
+// Method Description:
+// - If there is a pending update, tell it to not update the scroll bar's
+//   value (but still update the minimum, maximum and viewport size).
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void ScrollBarUpdater::CancelPendingValueChange()
+{
+    std::lock_guard guard(_pendingUpdateMutex);
+
+    if (_pendingUpdate.has_value())
+    {
+        _cancelledValueChange = true;
+    }
+}
+
+winrt::fire_and_forget ScrollBarUpdater::_RunUpdater()
+{
+    std::weak_ptr weakThis{ _stillExists };
+
+    // No need to lock because there should always be at most one `_RunUpdater`
+    // running at a time.
+    const auto waitTime = _throttle.GetNextWaitTime();
+    if (waitTime.has_value())
+    {
+        co_await winrt::resume_after(waitTime.value());
+
+        // `this` might have been destroyed at this point.
+        if (weakThis.expired())
+        {
+            co_return;
+        }
+    }
+
+    {
+        auto scrollBar = _scrollBar.get();
+        if (!scrollBar)
+        {
+            co_return;
+        }
+
+        co_await winrt::resume_foreground(scrollBar.Dispatcher());
+    }
+
+    // `this` might have been destroyed at this point.
+    if (weakThis.expired())

```

Comment:
This is actually wrong because I'm assuming that if it hasn't expired now, it will still be alive in the lines below but it's wrong because it might be destroyed later on the main thread.

I need to fix this (not ready for merge yet!)